crates/env/src/topics.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Docs
+
+use crate::EnvTypes;
+
+/// The concrete implementation that is guided by the topics builder.
+///
+/// To be implemented by the on-chain and off-chain environments respectively.
+#[doc(hidden)]
+pub trait TopicsBuilderBackend<E>
+where
+    E: EnvTypes,
+{
+    /// The type of the serialized event topics.
+    type Output;
+
+    /// Initialized the backend with the expected number of event topics.
+    fn expect(&mut self, expected_topics: usize);
+
+    /// Pushes another topic for serialization to the backend.
+    fn push_topic<T>(&mut self, topic_value: &T)
+    where
+        T: scale::Encode;
+
+    /// Extracts the serialized topics.
+    fn output(self) -> Self::Output;
+}
+
+/// Builder for event topic serialization.
+///
+/// Abstraction to build up event topic serialization with zero-overhead,
+/// no heap-memory allocations and no dynamic dispatch.
+#[doc(hidden)]
+pub struct TopicsBuilder<S, E, B> {
+    backend: B,
+    state: core::marker::PhantomData<fn() -> (S, E)>,
+}
+
+impl<E, B> From<B> for TopicsBuilder<state::Uninit, E, B>
+where
+    E: EnvTypes,
+    B: TopicsBuilderBackend<E>,
+{
+    fn from(backend: B) -> Self {
+        Self {
+            backend,
+            state: Default::default(),
+        }
+    }
+}
+
+#[doc(hidden)]
+pub mod state {
+    /// The topic builder is uninitialized and needs to be provided with the
+    /// expected number of topics that need to be constructed.
+    pub enum Uninit {}
```

Comment:
Just a matter of taste. I find those empty braces ugly.

```suggestion
    pub struct Uninit;
```

Reply:
There is a semantics difference. You can have instances of `struct Uninit;` but you cannot have live instances of `enum Uninit {}`. So if you want to use those types on a purely type level it is cleaner to have them as empty enums.

Reply:
I see. Makes sense.